### ISSUE TITLE
XGBoost training stage will not retry after one training task failed

### ISSUE THREAD

**Issue by jinmfeng** [dmlc/xgboost#8336]
We are training XGBoost model on spark, but our jobs frequently fails becuase of errors like node maintenance, resource preemption, disk usage issues. 
We know that for normal spark jobs if a task failed it will retry and be assigned to other normal nodes. But for XGBoost it's running in Barrier execution mode and the training step is trigger by collect() function, which means the stage is a ResultStage and a failed task will abort the whole stage. 
Our cluster is very big and busy, such failure are very common. So our training jobs fails freqeuntly when we train on a big dataset or train many models in the same application. 
Can you fix this by making the training stage to a ShuffleMapStage so that failed stage can auto retry? 

Our failed stages didn't retry:
<img width="1782" alt="image" src="https://user-images.githubusercontent.com/3907763/195539040-c5b2a1ae-a71a-4534-ad71-1db773a0222e.png">

DAGScheduler source code shows that the failed Stage will not retry if the stage is ResultStage and task is barrier task:
<img width="896" alt="image" src="https://user-images.githubusercontent.com/3907763/195542204-e137dec8-5dfe-410f-97a3-bf22eee3665c.png">

**Comment by trivialfis** [comment]
Closed by https://github.com/dmlc/xgboost/pull/9423 .
